src/core/state/backendNetworks/backendNetworks.ts

```diff
@@ -150,6 +153,78 @@ function createParameterizedSelector<T, Args extends unknown[]>(
   };
 }
 
+/**
+ * Get the diff of userChains and backend networks and append any missing new chains to userChains.
+ * Also update the userChainsOrder to put the new chains at the end of the array.
+ *
+ * @param networks - backend driven networks data
+ * @returns void
+ */
+const handleUpdateUserChains = (networks: BackendNetwork[]) => {
+  const { userChains, addUserChain } = useUserChainsStore.getState();
+
+  const newChains = networks.filter((network) => !userChains[+network.id]);
+  // if there are no new chains, then we don't need to update anything
+  if (!newChains.length) return;
+
+  newChains.forEach((chain) => addUserChain({ chainId: +chain.id }));
+};
+
+const handleDiffAndMergeRainbowChains = (networks: BackendNetwork[]) => {
+  const { rainbowChains, addCustomRPC, updateCustomRPC, setActiveRPC } =
+    useRainbowChainsStore.getState();
+
+  for (const newChain of networks) {
+    const existingChain = rainbowChains[+newChain.id];
+
+    const chain: Chain = {
+      id: +newChain.id,
+      nativeCurrency: {
+        name: newChain.nativeAsset.name,
+        symbol: newChain.nativeAsset.symbol,
+        decimals: newChain.nativeAsset.decimals,
+      },
+      name: newChain.name,
+      rpcUrls: {
+        default: {
+          http: [newChain.defaultRPC.url],
+        },
+        public: {
+          http: [newChain.defaultRPC.url],
+        },
+      },
+      blockExplorers: {
+        default: {
+          url: newChain.defaultExplorer.url,
+          name: newChain.defaultExplorer.label,
+        },
+      },
+    };
+
+    // if we don't have an existing chain, then we can simply add a new rainbow chain
+    if (
+      !existingChain ||
+      (existingChain &&
+        !existingChain.chains.find(
+          (chain) => chain.rpcUrls.default.http[0] === newChain.defaultRPC.url,
+        ))
+    ) {
+      addCustomRPC({ chain });
+      setActiveRPC({
+        chainId: +chain.id,
+        rpcUrl: chain.rpcUrls.default.http[0],
+      });
+      return;
+    }
+
+    // if we have an existing chain, then we update the existing chain
+    // and set the active RPC url to the backend network provided RPC url
+    updateCustomRPC({ chain });
+    setActiveRPC({ chainId: +chain.id, rpcUrl: chain.rpcUrls.default.http[0] });
+    return;
```

Comment:
I think we'd only want to call `setActiveRPC` to update a custom chain that becomes a supported chain (old public RPC -> Rainbow proxy RPC). This currently catches the scenario where a user has previously switched the default RPC for a supported chain from Rainbow to i.e. Flashbots, so we wouldn't want to override that preference